cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,282 @@
+      }
+    }
+    RETURN_NOT_OK(bool_builder_.FinishInternal(&out));
+    return Status::OK();
+  }
+
+ private:
+  BooleanBuilder bool_builder_;
+};
+
+// ----------------------------------------------------------------------
+// Kernel wrapper for generic hash table kernels
+
+template <typename Type, typename Enable = void>
+struct HashBinaryKernelTraits {};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_null<Type>> {
+  using HashBinaryKernelImpl = NullIsInKernel;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_has_c_type<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, typename Type::c_type>;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_boolean<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, bool>;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_binary<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, util::string_view>;
+};
+
+template <typename Type>
+struct HashBinaryKernelTraits<Type, enable_if_fixed_size_binary<Type>> {
+  using HashBinaryKernelImpl = IsInKernel<Type, util::string_view>;
+};
+
+Status GetIsInKernel(FunctionContext* ctx, const std::shared_ptr<DataType>& type,
+                     std::unique_ptr<HashBinaryKernelImpl>* out) {
+  std::unique_ptr<HashBinaryKernelImpl> kernel;
+
+#define ISIN_CASE(InType)                                                           \
+  case InType::type_id:                                                             \
+    kernel.reset(new typename HashBinaryKernelTraits<InType>::HashBinaryKernelImpl( \
+        type, ctx->memory_pool()));                                                 \
+    break
+
+  switch (type->id()) {
+    ISIN_CASE(NullType);
+    ISIN_CASE(BooleanType);
+    ISIN_CASE(UInt8Type);
+    ISIN_CASE(Int8Type);
+    ISIN_CASE(UInt16Type);
+    ISIN_CASE(Int16Type);
+    ISIN_CASE(UInt32Type);
+    ISIN_CASE(Int32Type);
+    ISIN_CASE(UInt64Type);
+    ISIN_CASE(Int64Type);
+    ISIN_CASE(FloatType);
+    ISIN_CASE(DoubleType);
+    ISIN_CASE(Date32Type);
+    ISIN_CASE(Date64Type);
+    ISIN_CASE(Time32Type);
+    ISIN_CASE(Time64Type);
+    ISIN_CASE(TimestampType);
+    ISIN_CASE(BinaryType);
+    ISIN_CASE(StringType);
+    ISIN_CASE(FixedSizeBinaryType);
+    ISIN_CASE(Decimal128Type);
+    default:
+      break;
+  }
+#undef ISIN_CASE
+
+  CHECK_IMPLEMENTED(kernel, "isin", type);
+  *out = std::move(kernel);
+  return Status::OK();
+}
+
+Status IsIn(FunctionContext* context, const Datum& left, const Datum& right, Datum* out) {
+  DCHECK(left.type()->Equals(right.type()));
+  std::unique_ptr<HashBinaryKernelImpl> isin_kernel;
+  RETURN_NOT_OK(GetIsInKernel(context, right.type(), &isin_kernel));
+
+  detail::PrimitiveAllocatingBinaryKernel kernel(isin_kernel.get());
+  out->value = ArrayData::Make(isin_kernel->out_type(), (left.make_array())->length());
+  kernel.Call(context, left, right, out);
```

Comment:
I think we should be using InvokeUnaryKernel to handle both cases of ARRAY_DATA and chunked arrays.  It looks like some of the other in kernels doesn't do this as well.  I'm not sure why that is @fsaintjacques?

An example of doing this for hash related functions is at https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/hash.cc#L490  note this could probably be cleaner but you should look how the use is slightly different between Unique and Counts

Reply:
If we want proper [multithread execution](https://docs.google.com/document/d/10RoUZmiMQRi_J1FcPeVAUAMJ6d_ZuiEbaM2Y33sNPu4/edit#), I believe the tasks needs to be exposed to the caller. Some kernels requires a reduce phase like aggregates, so the current Invoke*Kernel current implementation don't work (in the current form).

On the other side, the support of ChunkedArray make it more convenient for the wrapped kernel function (Cast, Count, Mean, ...).

Reply:
Right, I think the kernel will ultimately need to be exposed to the caller, but I think these calls are simple wrappers that are meant to handle as many cases as possible.